### PR TITLE
executor: fix bug for virtual column in transaction

### PR DESCRIPTION
**Pull request by Deardrops**
### What problem does this PR solve? 

Continue unfinished works in #13591

### What is changed and how it works?

When we exec `UnoinScan` executor, we should fill virtual column for avoiding `NULL` value returned.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Fix bugs when select from generated virtual columns in a transaction.

### CODE REVIEW

File: executor/table_reader.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ package executor
 import (
 	"context"
 	"fmt"
+	"github.com/pingcap/tidb/expression"
```
Comment (lonng) [inline executor/table_reader.go]: Sort the imports.

File: executor/union_scan.go [hunk 2]
```diff
@@ -156,7 +167,28 @@ func (us *UnionScanExec) Next(ctx context.Context, req *chunk.Chunk) error {
...
+			// Because the expression might return different type from
+			// the generated column, we should wrap a CAST on the result.
+			castDatum, err := table.CastValue(us.ctx, datum, us.columns[idx])
+			if err != nil {
+				return err
+			}
+			mutableRow.SetDatum(idx, castDatum)
+		}
+
+		matched, _, err := expression.EvalBool(us.ctx, us.conditionsWithVirCol, mutableRow.ToRow())
```
Comment (AilinKid) [inline executor/union_scan.go]: why need this？Is it possible that `matched` is false and `err` is nil?

Comment (Deardrops) [reply executor/union_scan.go]: This is a `selection` operator, used for `select a from t where a > 10`.
If one row has `a = 11`, this function will return `true`, if `a = 9`, return `false`.
We should filter all row meet the conditions, and discharge others rows.

**Pull request merged**